KAFKA-16913 - Support external schemas in JSONConverter

priyankaku opened the pull request:
When using a connector that requires a schema, such as JDBC connectors,
with JSON messages, the current JSONConverter necessitates including the
schema within every message. To address this, we are introducing a new
parameter, schema.content, which allows you to provide the schema
externally. This approach not only reduces the size of the messages but
also facilitates the use of more complex schemas.

KIP :
https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverter

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

Commit:
feat: Added external schema support for connector

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverte

Signed-off-by: Priyanka K U <priyanka.k.u@ibm.com>

FrankYang0529 commented on connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java:
```diff
@@ -978,6 +978,39 @@ public void testVersionRetrievedFromAppInfoParser() {
+
+    @Test
+    public void testSchemaContentIsEmptyString() {
+        converter.configure(Collections.singletonMap(JsonConverterConfig.SCHEMA_CONTENT_CONFIG, ""), false);
```
nit: `Collections.singletonMap` -> `Map.of`.

priyankaku replied:
Changes made as suggested.

FrankYang0529 commented on connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java:
```diff
@@ -978,6 +978,39 @@ public void testVersionRetrievedFromAppInfoParser() {
+
+    @Test
+    public void testSchemaContentValidSchema() {
+        converter.configure(Collections.singletonMap(JsonConverterConfig.SCHEMA_CONTENT_CONFIG, "{ \"type\": \"string\" }"), false);
```
ditto

priyankaku replied:
Comments addressed.

FrankYang0529 commented on connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java:
```diff
@@ -978,6 +978,39 @@ public void testVersionRetrievedFromAppInfoParser() {
+    public void testSchemaContentInValidSchema() {
+        assertThrows(
+            DataException.class,
+            () -> converter.configure(Collections.singletonMap(JsonConverterConfig.SCHEMA_CONTENT_CONFIG, "{ \"string\" }"), false),
```
ditto

priyankaku replied:
Comments addressed.

FrankYang0529 commented on connect/json/src/test/java/org/apache/kafka/connect/json/JsonConverterTest.java:
```diff
@@ -978,6 +978,39 @@ public void testVersionRetrievedFromAppInfoParser() {
+
+    @Test
+    public void testSchemaContentLooksLikeSchema() {
+        converter.configure(Collections.singletonMap(JsonConverterConfig.SCHEMA_CONTENT_CONFIG, "{ \"type\": \"struct\", \"fields\": [{\"field\": \"schema\", \"type\": \"struct\",\"fields\": [{\"field\": \"type\", \"type\": \"string\" }]}, {\"field\": \"payload\", \"type\": \"string\"}]}"), false);
```
ditto

FrankYang0529 reviewed (commented):
Thanks for the PR. I cannot find test case about `DataException` in `toConnectData`. Could you also add following case? Thank you.

```java
    @ParameterizedTest
    @ValueSource(strings = {
        "{ }",
        "{ \"wrong\": \"schema\" }",
        "{ \"schema\": { \"type\": \"string\" } }",
        "{ \"payload\": \"foo-bar-baz\" }",
        "{ \"schema\": { \"type\": \"string\" }, \"payload\": \"foo-bar-baz\", \"extra\": \"field\" }",
    })
    public void testNullSchemaContentWithWrongConnectDataValue(String value) {
        converter.configure(Map.of(), false);
        assertThrows(
            DataException.class,
            () -> converter.toConnectData(TOPIC, value.getBytes()));
    }
```

Commit:
feat:  addressing review comments

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverte

Signed-off-by: Priyanka K U <priyanka.k.u@ibm.com>

priyankaku committed:
feat: addressing review comments

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverte

Signed-off-by: Priyanka K U <priyanka.k.u@ibm.com>

priyankaku commented:
> Thanks for the PR. I cannot find test case about `DataException` in `toConnectData`. Could you also add following case? Thank you.
> 
> ```java
>     @ParameterizedTest
>     @ValueSource(strings = {
>         "{ }",
>         "{ \"wrong\": \"schema\" }",
>         "{ \"schema\": { \"type\": \"string\" } }",
>         "{ \"payload\": \"foo-bar-baz\" }",
>         "{ \"schema\": { \"type\": \"string\" }, \"payload\": \"foo-bar-baz\", \"extra\": \"field\" }",
>     })
>     public void testNullSchemaContentWithWrongConnectDataValue(String value) {
>         converter.configure(Map.of(), false);
>         assertThrows(
>             DataException.class,
>             () -> converter.toConnectData(TOPIC, value.getBytes()));
>     }
> ```

I have added this Test case, thanks for suggesting.

priyankaku commented:
Hi, @FrankYang0529 
Thanks for your Feedback, I have  incorporated the changes as suggested , Please take a look.

priyankaku commented:
@FrankYang0529  : If you get sometime , Could you please re review the PR?

priyankaku commented:
@chia7712 , it would great if you could review this PR

priyankaku commented:
@chia7712 : Apologies for the repeated follow ups. I kindly request your review on this when you have a moment.

edoardocomar commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverterConfig.java:
```diff
@@ -130,4 +140,17 @@ public boolean replaceNullWithDefault() {
+        return schemaContent;
+    }
+
+
```
nit: no need for these two whitespace lines

priyankaku replied:
Addressed

edoardocomar commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverterConfig.java:
```diff
@@ -61,6 +66,8 @@ public final class JsonConverterConfig extends ConverterConfig {
                       orderInGroup++, Width.MEDIUM, SCHEMAS_CACHE_SIZE_DISPLAY);
+        CONFIG.define(SCHEMA_CONTENT_CONFIG, Type.STRING, SCHEMA_CONTENT_DEFAULT, Importance.HIGH, SCHEMA_CONTENT_DOC, group, 
+                      orderInGroup++, Width.MEDIUM, SCHEMA_CONTENT_DISPLAY);
 
```
Could you have called the overloaded `define` which uses a Validator, for a minimal check that the value is ok ?
As in the JSON converter for example you check not null and not 0 length.

priyankaku replied:
Hi @edoardocomar  Thanks for suggestion.
However, do we really need to include a validator here Since we are setting  SCHEMA_CONTENT_CONFIG  as an optional field which defaults to null. Please correct me if am wrong.

Commit:
feat: addressing review comments

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverte

mimaison commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverterConfig.java:
```diff
@@ -36,6 +37,10 @@ public final class JsonConverterConfig extends ConverterConfig {
     private static final String SCHEMAS_ENABLE_DISPLAY = "Enable Schemas";
 
     public static final String SCHEMAS_CACHE_SIZE_CONFIG = "schemas.cache.size";
+    public static final String SCHEMA_CONTENT_CONFIG = "schema.content";
```
Can we keep the `schemas.cache.size` fields together instead of adding these new fields in the middle?

priyankaku replied:
Addressed.

mimaison commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -291,6 +295,17 @@ public void configure(Map<String, ?> configs) {
+        } catch (SerializationException e) {
+            throw new DataException("Failed to parse schema in converter config due to serialization error: ", e);
+        }
+
```
We don't need this extra line

priyankaku replied:
Addressed.

mimaison commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -345,13 +360,16 @@ public SchemaAndValue toConnectData(String topic, byte[] value) {
-        if (config.schemasEnabled() && (!jsonValue.isObject() || jsonValue.size() != 2 || !jsonValue.has(JsonSchema.ENVELOPE_SCHEMA_FIELD_NAME) || !jsonValue.has(JsonSchema.ENVELOPE_PAYLOAD_FIELD_NAME)))
-            throw new DataException("JsonConverter with schemas.enable requires \"schema\" and \"payload\" fields and may not contain additional fields." +
+        if (config.schemasEnabled()) {
+            if (this.schema != null) {
```
We can just use `schema` here like you did just below

priyankaku replied:
Addressed.

mimaison commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -68,6 +68,10 @@ public class JsonConverter implements Converter, HeaderConverter, Versioned {
 
+    // if a schema is provided in config, this schema will 
+    // be used for all messages
+    private Schema schema = null;
```
Can we put this with the other non-static, non-final fields below?

priyankaku replied:
Addressed.

mimaison commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -68,6 +68,10 @@ public class JsonConverter implements Converter, HeaderConverter, Versioned {
     private static final Map<Schema.Type, JsonToConnectTypeConverter> TO_CONNECT_CONVERTERS = new EnumMap<>(Schema.Type.class);
 
+    // if a schema is provided in config, this schema will 
+    // be used for all messages
```
A short comment like this can fit on a single line

priyankaku replied:
Addressed.

mimaison commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -291,6 +295,17 @@ public void configure(Map<String, ?> configs) {
+
+        try {
+            final byte[] schemaContent = config.schemaContent();
+            if (schemaContent != null && schemaContent.length > 0) {
```
I wonder if the empty string case should be handled in JsonConverterConfig, WDYT?

priyankaku replied:
I have addressed  this , could you please take a look and if this is what you expected.

mimaison reviewed (commented):
Thanks for the PR. I left a few comments.

Commit:
feat: addressing review comments

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverter

priyankaku commented:
@mimaison : Thank you for reviewing this , I have addressed the comments . Please let me know .

mimaison commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -68,6 +68,8 @@ public class JsonConverter implements Converter, HeaderConverter, Versioned {
 
     private static final Map<Schema.Type, JsonToConnectTypeConverter> TO_CONNECT_CONVERTERS = new EnumMap<>(Schema.Type.class);
 
+    // if a schema is provided in config, this schema will be used for all messages
```
This comment be next to the new field

priyankaku replied:
@mimaison : sorry I missed this out , made the changes. please take a look

Commit:
feat: addressed review comments

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverter

priyankaku commented:
@mimaison : Hi Mickael , I have addressed all the comments you had suggested so far. Could you re review again if you some time.

mimaison commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverterConfig.java:
```diff
@@ -35,6 +36,11 @@ public final class JsonConverterConfig extends ConverterConfig {
 
+    public static final String SCHEMA_CONTENT_CONFIG = "schema.content";
+    public static final String SCHEMA_CONTENT_DEFAULT = null;
+    private static final String SCHEMA_CONTENT_DOC = "When set, this is used as the schema for all messages. Otherwise, the schema will should be in the contents of each message.";
```
One last nit: The 2nd sentence does not seem right, it should be either `the schema will be` or `the schema should be`

priyankaku replied:
Thanks for pointing this out , I updated the message.

Commit:
feat: fixed review comments

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverter

Commit:
Merge remote-tracking branch 'upstream/trunk' into KIP-1054-external-schema-support

Commit:
feat: updated Collections to Map

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverter

Commit:
feat: fix for  unit test case

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverter

Commit:
feat: updated test case

Contributes to: event-integration/eventstreams-planning/12766
KIP Link:  https://cwiki.apache.org/confluence/display/KAFKA/KIP-1054%3A+Support+external+schemas+in+JSONConverter

mimaison reviewed (approved):
Thanks for the updates, LGTM

mimaison commented:
@frankvicky @edoardocomar Do you have any other comments?

chia7712 commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverterConfig.java:
```diff
@@ -35,6 +36,11 @@ public final class JsonConverterConfig extends ConverterConfig {
 
+    public static final String SCHEMA_CONTENT_CONFIG = "schema.content";
+    public static final String SCHEMA_CONTENT_DEFAULT = null;
+    private static final String SCHEMA_CONTENT_DOC = "When set, this is used as the schema for all messages. Otherwise, the schema will be included in the content of each message.";
```
The configuration is used for the sink connector, not the source connector. Is it correct? If so, could you please add it to the documentation?

priyankaku replied:
yes this is specific to sink connector , I will be adding this to documentation.

chia7712 commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -340,13 +351,16 @@ public SchemaAndValue toConnectData(String topic, byte[] value) {
-            throw new DataException("JsonConverter with schemas.enable requires \"schema\" and \"payload\" fields and may not contain additional fields." +
+        if (config.schemasEnabled()) {
+            if (schema != null) {
+                return new SchemaAndValue(schema, convertToConnect(schema, jsonValue, config));
```
What happens if the message has the schema field?

priyankaku replied:
If schema.enable is set to true and schema.content has a value, only the value from schema.content will be used. Schema in the message payload will be ignored

chia7712 replied:
If `jsonValue` contains the `schema` and `payload` fields, should we call `jsonValue.get(JsonSchema.ENVELOPE_PAYLOAD_FIELD_NAME)` instead? Or is this case - where `jsonValue` includes the schema - not supported by `schema.content`?

JimmyWang6 replied:
Hi @chia7712, I just noticed that if `jsonValue` contains valid schema and payload fields and `schema.content` is null, current logic will go down to the end of this method:
```java
Schema schema = asConnectSchema(jsonValue.get(JsonSchema.ENVELOPE_SCHEMA_FIELD_NAME));
        return new SchemaAndValue(
                schema,
                convertToConnect(schema, jsonValue.get(JsonSchema.ENVELOPE_PAYLOAD_FIELD_NAME), config)
        );
``` 
In this case, the method uses the corresponding schemas within each of the serialized values and keys, which I think maybe it's the expected behavior. What do you think?

chia7712 replied:
> In this case, the method uses the corresponding schemas within each of the serialized values and keys, which I think maybe it's the expected behavior. What do you think?

yes, that is expected behavior.

The scenario I mentioned was `schema.content` is NOT null and the `jsonValue` has both fields `payload` and `schema`

chia7712 commented on connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java:
```diff
@@ -340,13 +351,16 @@ public SchemaAndValue toConnectData(String topic, byte[] value) {
-        if (config.schemasEnabled() && (!jsonValue.isObject() || jsonValue.size() != 2 || !jsonValue.has(JsonSchema.ENVELOPE_SCHEMA_FIELD_NAME) || !jsonValue.has(JsonSchema.ENVELOPE_PAYLOAD_FIELD_NAME)))
-            throw new DataException("JsonConverter with schemas.enable requires \"schema\" and \"payload\" fields and may not contain additional fields." +
+        if (config.schemasEnabled()) {
+            if (schema != null) {
```
Users need to set `schemas.enable=false` in the source connector and then set `schemas.enable=true` along with `schema.content=xxx` in the sink connector to avoid embedding the schema in every message. Is this understanding correct? If so, the `schema.enable` setting might be confusing, as the same flag has completely different meanings for the source and sink connectors

priyankaku replied:
Yes your understanding is right , we have extended the purpose of the schema.enable flag by adding an additional capability, and I don’t believe this will cause significant confusion.
To avoid any further confusion we can provide a detailed description in documentation explaining how this flag behaves in both scenarios , when schema.enable is set and when it is not.

chia7712 replied:
Yes, it would be great if you could improve the use case by updating the documentation

chia7712 reviewed (commented):
@priyankaku Sorry for the late review. I have a couple of questions. Please take a look. thanks!
